src/agentlab2/core.py

```diff
@@ -0,0 +1,118 @@
+import base64
+import io
+from typing import Any, Callable, Dict, List, Optional, Self
+
+import litellm.utils
+from litellm import Message
+from PIL import Image
+from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator
+
+
+class ActionSchema(BaseModel):
+    """
+    Represents a function specification with a type, name, description and arguments.
+    Compatible with OAI, Anthropic and VLLM definitions.
+
+    Attributes:
+        type (Literal["function"]): The type of the tool, which is always "function".
+        name (str): The name of the function.
+        description (str): A brief description of the function.
+        parameters (dict): A dictionary containing the parameters of the function.
+    """
+
+    name: str
+    description: str
+    parameters: dict = Field(default_factory=dict)
+
+    @classmethod
+    def from_function(cls, func: Callable) -> Self:
+        """Create tool object from python function."""
+        schema = litellm.utils.function_to_dict(func)
+        return cls(**schema)
+
+    def schema(self) -> dict[str, Any]:
+        """Produce dict that could be passed as tool schema into LLM api."""
+        return {
+            "type": "function",
+            "function": {
+                "name": self.name,
+                "description": self.description,
+                "parameters": self.parameters,
+            },
+        }
+
+
+class Action(BaseModel):
+    """
+    A class representing a function call.
+
+    Attributes:
+        id (str): The identifier for the tool call.
+        name (str): The name of the function being called.
+        arguments (Any): The arguments to be passed to the function.
+    """
+
+    id: str | None = None
+    name: str
+    arguments: Dict[str, Any] = Field(default_factory=dict)
+
+
+image_prefix = "data:image/png;base64,"
+
+
+class Content(BaseModel):
+    """Represents a piece of content in an observation."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
```

Comment:
what is model_config?

Reply:
https://docs.pydantic.dev/2.0/usage/model_config/

it's better than passing random kw values in the class def